pkg/prometheus/promcfg_test.go

```diff
@@ -8783,3 +8783,205 @@ func TestPodMonitorWithNonDefaultScrapeClassRelabelings(t *testing.T) {
 	require.NoError(t, err)
 	golden.Assert(t, string(cfg), "podMonitorObjectWithNonDefaultScrapeClassWithRelabelings.golden")
 }
+
+func TestServiceMonitorWithDefaultScrapeClassMetricRelabelings(t *testing.T) {
+	prometheus := defaultPrometheus()
+	prometheus.Spec.CommonPrometheusFields.EnforcedNamespaceLabel = "namespace"
+	serviceMonitor := defaultServiceMonitor()
+	scrapeClasses := []monitoringv1.ScrapeClass{
+		{
+			Name:    "default",
+			Default: ptr.To(true),
+			MetricRelabelings: []monitoringv1.RelabelConfig{
+				{
+					SourceLabels: []monitoringv1.LabelName{"namespace"},
+					Regex:        "tenant1-.*",
+					TargetLabel:  "tenant",
+					Replacement:  ptr.To("tenant1"),
+				},
+				{
+					SourceLabels: []monitoringv1.LabelName{"namespace"},
+					Regex:        "tenant2-.*",
+					TargetLabel:  "tenant",
+					Replacement:  ptr.To("tenant2"),
+				},
+			},
+		},
+		{
+			Name: "not-default",
+			MetricRelabelings: []monitoringv1.RelabelConfig{
+				{
+					TargetLabel: "tenant",
+					Replacement: ptr.To("not-default"),
+				},
+			},
+		},
+	}
+
+	prometheus.Spec.ScrapeClasses = scrapeClasses
+	cg := mustNewConfigGenerator(t, prometheus)
+
+	cfg, err := cg.GenerateServerConfiguration(
+		context.Background(),
+		prometheus.Spec.EvaluationInterval,
+		prometheus.Spec.QueryLogFile,
+		prometheus.Spec.RuleSelector,
+		prometheus.Spec.Exemplars,
+		prometheus.Spec.TSDB,
+		prometheus.Spec.Alerting,
+		prometheus.Spec.RemoteRead,
+		map[string]*monitoringv1.ServiceMonitor{"monitor": serviceMonitor},
+		nil,
+		nil,
+		nil,
+		&assets.StoreBuilder{},
+		nil,
+		nil,
+		nil,
+		nil,
+	)
+	require.NoError(t, err)
+	golden.Assert(t, string(cfg), "serviceMonitorObjectWithDefaultScrapeClassWithMetricRelabelings.golden")
```

Comment:
What I mean with code duplication is that this part is basically identical in all tests. We could have a single function with different inputs instead

```go
testCases := []struct{
  scrapeClass monitoringv1.ScrapeClass
  podMonitors map[string]monitoringv1.Podmonitor // can be nil
  serviceMonitors map[string]monitoringv1.ServiceMonitor // can be nil
  probes map[string]monitoringv1.Probe // can be nil
  scrapeConfigs map[string]monitoringv1alpha1.ScrapeConfig // can be nil
  goldenFile string  
}
```
You can put the implementation of the test inside a loop and asser that your inputs results are equal to the appropriate golden file